src/PrestaShopBundle/Controller/Admin/Improve/Design/MailThemeController.php

```diff
@@ -0,0 +1,378 @@
+
+        return new Response($renderedLayout);
+    }
+
+    /**
+     * Display the raw source of a theme layout (mainly useful for developers/integrators)
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
+     * @param string $theme
+     * @param string $layout
+     * @param string $type
+     * @param string $locale
+     * @param string $module
+     *
+     * @return Response
+     *
+     * @throws FileNotFoundException
+     * @throws InvalidArgumentException
+     */
+    public function rawLayoutAction($theme, $layout, $type, $locale, $module = '')
+    {
+        $renderedLayout = $this->renderLayout($theme, $layout, $type, $locale, $module);
+
+        $response = new Response($renderedLayout, 200, [
+            'Content-Type' => 'text/plain',
+        ]);
+
+        return $response;
+    }
+
+    /**
+     * Dynamically display an email template, this is usually used by the MailGenerator but this action
+     * allows to display preview before generating the file (handy when you are developing an email theme)
+     *
+     * @param string $themeName
+     * @param string $layoutName
+     * @param string $type
+     * @param string $locale
+     * @param string $module
+     *
+     * @return string
+     *
+     * @throws FileNotFoundException
+     * @throws InvalidArgumentException
+     */
+    private function renderLayout($themeName, $layoutName, $type, $locale = '', $module = '')
+    {
+        /** @var ThemeCatalogInterface $themeCatalog */
+        $themeCatalog = $this->get('prestashop.core.mail_template.theme_catalog');
+        /** @var ThemeInterface $theme */
+        $theme = $themeCatalog->getByName($themeName);
+
+        /** @var LayoutInterface $layout */
+        $layout = null;
+        /* @var LayoutInterface $layoutInterface */
+        foreach ($theme->getLayouts() as $layoutInterface) {
+            if ($layoutInterface->getName() == $layoutName
+                && $layoutInterface->getModuleName() == $module
+            ) {
+                $layout = $layoutInterface;
+                break;
+            }
+        }
+
+        if (null === $layout) {
+            throw new FileNotFoundException(sprintf(
+                'Could not find layout %s%s in theme %s',
+                empty($module) ? '' : $module . ':',
+                $layoutName,
+                $themeName
+            ));
+        }
+
+        /** @var LanguageRepositoryInterface $languageRepository */
+        $languageRepository = $this->get('prestashop.core.admin.lang.repository');
+        if (empty($locale)) {
+            $locale = $this->getContext()->language->locale;
+        }
+        /** @var LanguageInterface $language */
+        $language = $languageRepository->getByLocaleOrIsoCode($locale);
+        if (null === $language) {
+            throw new InvalidArgumentException(sprintf('Could not find Language with locale or isoCode %s', $locale));
```

Comment:
Same question here. Let me know. :-)

Reply:
Same here, only happen if someone tries to enter an invalid url which is not supposed to happen